### ISSUE TITLE
Loss of precision in Core::toString()

### ISSUE THREAD

**Issue by andres-h** [SeisComP/common#73]
Core::toString() has fixed precision of 10, which means that, eg., double loses precision when serialized as XML. We cannot be sure that 10 digits precision is sufficient for all use cases.

https://github.com/SeisComP/common/blob/348311ade9a713d92b301d672ee54e71622c6e60/libs/seiscomp/core/strings.ipp#L50

Suggest using `ostream.precision(std::numeric_limits<T>::max_digits10)`

See: https://en.cppreference.com/w/cpp/types/numeric_limits/max_digits10

**Comment by gempa-jabe** [comment]
I agree that 10 was used to have a good compromise of accuracy and formatting. Here is a nice example of what happens with that modification:

```xml
<sensor publicID="Sensor#20120426173219.391931.23897" name="AAK.1990.285.HE00" response="ResponsePAZ#20120426173219.39257.23898">
  <description>STS-1H</description>
  <model>STS-1H</model>
  <manufacturer>Streckeisen</manufacturer>
  <type>VBB</type>
  <unit>M/S</unit>
  <calibration serialNumber="yyyy" channel="2">
    <start>1990-10-12T00:00:00.0000Z</start>
    <end>1991-04-16T23:59:59.0000Z</end>
    <gain>1053</gain>
    <gainFrequency>0.050000000000000003</gainFrequency>
  </calibration>
</sensor>
```

What do you think?

**Comment by andres-h** [comment]
I think preventing the loss of accuracy is more important than nice formatting. The concrete problem is that filter coefficients in StationXML are losing accuracy after import. Some filters are known to be rather unstable, so this may have noticeable impact.

10 decimal places is effectively the precision of float (rather than double).

**Comment by gempa-jabe** [comment]
10 was not chosen to match the highest precision possible but as a compromise as outlined above.

```
std::numeric_limits<float>::max_digits10 = 9
std::numeric_limits<double>::max_digits10 = 17
```

I can live with it although it will output lots of undesired number strings. See 8c59f27a9839b044aeea1568443dbd82325e9437 and try yourself.

**Comment by gempa-jabe** [comment]
Fixing the current tests of e.g. fdsnws I am not very happy with numbers like:

```
0.00050000000000000001
```

I have added a function which retrieves the most significant digits of a number and will only output those. I am going to measure performance and maybe that is then a better solution.

**Comment by gempa-jabe** [comment]
@andres-h can you give an example of numbers which are not printed with sufficient precision?

**Comment by gempa-jabe** [comment]
Any feedback here? Can we close the issue?

c1738d093144800d39e966cb1ccb4ec0ee121d8b implements an output with the most significant digits. Does that work for you?

**Comment by andres-h** [comment]
Thanks, looks OK to me now.

> @andres-h can you give an example of numbers which are not printed with sufficient precision?

Few examples:

Original StationXML: 608949.787164006
Old version: 608949.7872 (insufficient precision)
New version: 608949.787164006 (OK)

Original StationXML: -0.00027956301912657736
Old version: -0.0002795630191 (insufficient precision)
New version: -0.00027956301912657742 (small difference, but OK)

Original StationXML: 0.0005
Old version: 0.0005 (OK)
New version: 0.0005 (OK)

Original StationXML: 1987.0
Old version: 1987.0 (OK)
New version: 1987 (OK)

**Comment by gempa-jabe** [comment]
-0.00027956301912657736 should yield "-0.0002795630191265774", at least for me. I am going to check why it can't be output at full precision.

**Comment by gempa-jabe** [comment]
I pushed a fix, please check again if that is also working for you.

**Comment by andres-h** [comment]
It works for me, but the exact choice of precision is not fully obvious. Maybe using just (std::numeric_limits<T>::max_digits10 - 1), eg., 16, would be good enough?

Original StationXML: -0.00027956301912657736
New version 1: -0.00027956301912657742 (small difference, but OK)
New version 2: -0.00027956301912657736 (OK, precision (17))

Original StationXML: -0.035650661618199464
New version 1: -0.035650661618199464 (OK)
New version 2: -0.03565066161819946 (precision(16))

Original StationXML: 0.0001309998199072425
New version 1: 0.0001309998199072425 (OK)
New version 2: 0.00013099981990724249 (precision(17))

**Comment by gempa-jabe** [comment]
The algorithm tries to figure out the required significant digits. Always using 16 will lead to undesired numbers, e.g. 0.50000000001.

**Comment by gempa-jabe** [comment]
I have added tests for that and you can play around with them. The algorithm is not difficult and I think that it is producing pretty good formatting.

**Comment by andres-h** [comment]
> The algorithm tries to figure out the required significant digits. Always using 16 will lead to undesired numbers, e.g. 0.50000000001.

The drawback would be a small loss of precision, but I think formatting would be OK.

https://coliru.stacked-crooked.com/a/ecf2cd424decc6f3

**Comment by gempa-jabe** [comment]
Looks interesting. I am going to check that next week.

**Comment by gempa-jabe** [comment]
The code linked above would render `-0.00027956301912657736` as `-0.0002795630191265774` which is something else given the intended precision. Is also renders `0.9999999999999998` as `0.9999999999999998` and `0.50000000001` as `0.50000000001`. It also creates number like `0.07347049999999999` read from the database. This is unfortunately not a generic solution.

**Comment by jsaul** [comment]
I am not entirely sure what the goal is. Clearly you want to format 0.9999999999999998 as 1 or 1.0, and 0.50000000001 as 0.5. There may be code that "guesses" that the intention is to format 0.50000000001 as 0.5 even though there is no guarantee that this guess is correct.

But what if the number is e.g. sqrt(2.)? How would anyone guess what the intended output looks like in terms number of decimals?

@andres-h is right that limiting the precision is risky and should be avoided. And I think there is no generic solution other than optionally passing the desired precision (or a format string) to toString(). By default the full precision should always be preserved and where "pretty printing" is needed, the formatting should be requested explicitly. I do understand that depending on context this is a real challenge and may not be feasible at all.

**Comment by gempa-jabe** [comment]
> @andres-h is right that limiting the precision is risky and should be avoided.

We have been doing that for more than a decade now ;) So the risk has been pretty low I think.

The current implementation does not depend on the context. Past versions (the current release as well) limits the precision to 10. The current code "detects" the significant digits and improves the output already significantly. Preserving the full precision at any time is nothing that I support. Even limiting it to max-1 is a hack and will remove the last digit. If you really want to have full binary precision then we need to store numbers as binary e.g. base64 encoded. That might even look prettier than 0.50000000001.

Basically the current algorithms cuts everything of the fractional part following six zeros or six nines after a non zero digit.
